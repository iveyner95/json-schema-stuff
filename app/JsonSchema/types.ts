import { Node } from 'reactflow';

export type JsonSchema = Partial<Record<string, any>>;
export type JsonTraverseSchemaFn = (schema: JsonSchema, sourceNodeId: string) => void;
export type AddNodeFn = (label: string, jsonSchema: JsonSchema) => void;
export type AddEdgeFn = (target: string, source: string) => void;
export type GetLastNodeIdFn = () => string;
export type SubschemaExistsFn = (schema: JsonSchema, subschemaKey: string) => boolean;

export interface IJsonSchemaTraverser {
  traverseSchema: JsonTraverseSchemaFn;
}

export interface JsonSchemaTypeTraverserArgs {
  subschemaExists: SubschemaExistsFn;
  traverseSubschema: JsonTraverseSchemaFn;
}

export type NodeSchemaData = Record<Node['id'], JsonSchema>;

export enum JsonSchemaType {
  STRING = 'string',
  NUMBER = 'number',
  INTEGER = 'integer',
  OBJECT = 'object',
  ARRAY = 'array',
  BOOLEAN = 'boolean',
  NULL = 'null',
}
