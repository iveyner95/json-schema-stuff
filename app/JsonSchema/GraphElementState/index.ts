export * from './GraphElementState';
export * from './types';
