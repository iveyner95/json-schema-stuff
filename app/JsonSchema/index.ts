export * from './GraphElementState';
export * from './JsonSchemaTraversal';
export * from './JsonSchemaTraverserInitializer';
export * from './parseSchema';
export * from './types';
