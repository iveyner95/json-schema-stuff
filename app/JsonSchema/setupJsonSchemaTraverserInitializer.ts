'use client';

import { IGraphElementState } from './GraphElementState';
import { JsonSchemaTraverser } from './JsonSchemaTraversal';
import { JsonSchemaTraverserInitializer } from './JsonSchemaTraverserInitializer';

export const setupJsonSchemaTraverserInitializer = (
  graphElementState: IGraphElementState,
  schema?: JSON
) => {
  const jsonSchemaTraverser = new JsonSchemaTraverser(graphElementState);
  const jsonSchemaTraverserInitializer = new JsonSchemaTraverserInitializer(
    schema,
    graphElementState,
    jsonSchemaTraverser
  );

  return jsonSchemaTraverserInitializer;
};
