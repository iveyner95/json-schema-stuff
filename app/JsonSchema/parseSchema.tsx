'use client';

import { GraphElementState } from './GraphElementState';
import { setupJsonSchemaTraverserInitializer } from './setupJsonSchemaTraverserInitializer';

export function parseSchema(schema: JSON | undefined) {
  const graphElementState = new GraphElementState();
  const jsonSchemaTraverserInitializer = setupJsonSchemaTraverserInitializer(
    graphElementState,
    schema
  );

  jsonSchemaTraverserInitializer.parse();

  return {
    edges: graphElementState.edges,
    nodes: graphElementState.nodes,
    nodeSchemaData: graphElementState.nodeSchemaData,
  };
}
