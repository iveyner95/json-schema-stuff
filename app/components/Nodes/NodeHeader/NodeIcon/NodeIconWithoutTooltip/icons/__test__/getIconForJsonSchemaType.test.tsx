import { JsonSchemaType } from '../../../../../../../JsonSchema';
import { getIconForJsonSchemaType } from '../getIconForJsonSchemaType';

describe('getIconForJsonSchemaType', () => {
  const testCases: JsonSchemaType[] = [
    ...Object.values(JsonSchemaType),
    'notAJsonSchemaType' as JsonSchemaType,
  ];

  it.each(testCases)(
    'should return expected component when jsonSchemaType is %s',
    (jsonSchemaType: JsonSchemaType) => {
      const Icon = getIconForJsonSchemaType(jsonSchemaType);

      expect(Icon).toMatchSnapshot();
    }
  );
});
