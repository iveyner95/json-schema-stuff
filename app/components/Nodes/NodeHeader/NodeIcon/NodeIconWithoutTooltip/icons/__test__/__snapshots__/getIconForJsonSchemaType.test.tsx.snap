// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is array 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is boolean 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is integer 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is notAJsonSchemaType 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is null 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is number 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is object 1`] = `[Function]`;

exports[`getIconForJsonSchemaType should return expected component when jsonSchemaType is string 1`] = `[Function]`;
